compiler/describe/analyze.go

```diff
@@ -0,0 +1,160 @@
+package describe
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/brimdata/zed/compiler/ast/dag"
+	"github.com/brimdata/zed/compiler/data"
+	"github.com/brimdata/zed/compiler/optimizer"
+	"github.com/brimdata/zed/lake"
+	"github.com/brimdata/zed/order"
+	"github.com/brimdata/zed/pkg/field"
+	"github.com/segmentio/ksuid"
+)
+
+type Info struct {
+	Sources  []Source  `json:"sources"`
+	Channels []Channel `json:"channels"`
+}
+
+type Source interface {
+	Source()
+}
+
+type (
+	LakeMeta struct {
+		Kind string `json:"kind" unpack:""`
+		Meta string `json:"meta"`
+	}
+	Pool struct {
+		Kind string      `json:"kind" unpack:""`
+		Name string      `json:"name"`
+		ID   ksuid.KSUID `json:"id"`
+	}
+	Path struct {
+		Kind string `json:"kind" unpack:""`
+		URI  string `json:"uri"`
+	}
+)
+
+func (*LakeMeta) Source() {}
+func (*Pool) Source()     {}
+func (*Path) Source()     {}
+
+type Channel struct {
+	AggregationKeys field.List     `json:"aggregation_keys"`
+	Sort            *order.SortKey `json:"sort"`
+}
+
+func Analyze(ctx context.Context, source *data.Source, seq dag.Seq) (*Info, error) {
+	var info Info
+	var err error
+	if info.Sources, err = describeSources(ctx, source.Lake(), seq[0]); err != nil {
+		return nil, err
+	}
+	sortKeys, err := optimizer.New(ctx, source).SortKeys(seq)
+	if err != nil {
+		return nil, err
+	}
+	aggKeys := describeAggs(seq, []field.List{nil})
+	for i := range sortKeys {
+		// Convert SortKey to a pointer so a nil sort is encoded as null for
+		// JSON/ZSON.
+		var s *order.SortKey
+		if !sortKeys[i].IsNil() {
+			s = &sortKeys[i]
+		}
+		info.Channels = append(info.Channels, Channel{
+			Sort:            s,
+			AggregationKeys: aggKeys[i],
+		})
+	}
+	return &info, nil
+}
+
+func describeSources(ctx context.Context, lk *lake.Root, o dag.Op) ([]Source, error) {
+	switch o := o.(type) {
+	case *dag.Fork:
+		var s []Source
+		for _, p := range o.Paths {
+			out, err := describeSources(ctx, lk, p[0])
+			if err != nil {
+				return nil, err
+			}
+			s = append(s, out...)
+		}
+		return s, nil
+	case *dag.DefaultScan:
+		return []Source{&Path{Kind: "Path", URI: "stdio://stdin"}}, nil
+	case *dag.FileScan:
+		return []Source{&Path{Kind: "Path", URI: o.Path}}, nil
+	case *dag.HTTPScan:
+		return []Source{&Path{Kind: "Path", URI: o.URL}}, nil
+	case *dag.PoolScan:
+		return sourceOfPool(ctx, lk, o.ID)
+	case *dag.Lister:
+		return sourceOfPool(ctx, lk, o.Pool)
+	case *dag.SeqScan:
+		return sourceOfPool(ctx, lk, o.Pool)
+	case *dag.CommitMetaScan:
+		return sourceOfPool(ctx, lk, o.Pool)
+	case *dag.LakeMetaScan:
+		return []Source{&LakeMeta{Kind: "LakeMeta", Meta: o.Meta}}, nil
+	default:
+		return nil, fmt.Errorf("unsupported source type %T", o)
+	}
+}
+
+func sourceOfPool(ctx context.Context, lk *lake.Root, id ksuid.KSUID) ([]Source, error) {
+	if lk == nil {
+		panic(errors.New("internal error: lake operation cannot be used in non-lake context"))
```

Comment:
Don't panic when you can return an error.
```suggestion
		return nil, errors.New("internal error: lake operation cannot be used in non-lake context")
```

Reply:
We do a lot of returning errors in cases like this where this is clearly error that should have been caught in the semantic pass. My instinct is to stop returning these `internal errors: ...` and just panicking because if we get here it means something is seriously wrong. What do you think?

Reply:
I'm with you.

With that in mind then, how about removing this check and just letting the call to lk.OpenPool panic if lk is nil?

Reply:
Let's go!